domain/consensus/processes/blockprocessor/validateandinsertblock.go

```diff
@@ -179,6 +137,62 @@ func (bp *blockProcessor) validateAndInsertBlock(block *externalapi.DomainBlock)
 	return nil
 }
 
+func (bp *blockProcessor) validateAgainstSyncStateAndResolveInsertMode(block *externalapi.DomainBlock) (insertMode, error) {
+	syncInfo, err := bp.syncManager.GetSyncInfo()
+	if err != nil {
+		return 0, err
+	}
+	syncState := syncInfo.State
+
+	isHeaderOnlyBlock := isHeaderOnlyBlock(block)
+	blockHash := consensushashing.HeaderHash(block.Header)
+	if syncState == externalapi.SyncStateAwaitingGenesis {
+		if isHeaderOnlyBlock {
+			return 0, errors.Errorf("Got a header-only block while awaiting genesis")
+		}
+		if *blockHash != *bp.genesisHash {
+			return 0, errors.Errorf("Received a non-genesis block while awaiting genesis")
+		}
+		return insertModeGenesis, nil
+	}
+
+	if isHeaderOnlyBlock {
+		return insertModeHeader, nil
+	}
+
+	if syncState == externalapi.SyncStateAwaitingUTXOSet {
+		headerTipsPruningPoint, err := bp.consensusStateManager.HeaderTipsPruningPoint()
+		if err != nil {
+			return 0, err
+		}
+		if *blockHash != *headerTipsPruningPoint {
+			return 0, errors.Errorf("cannot insert blocks other than the header pruning point " +
+				"while awaiting the UTXO set")
+		}
+		return insertModeBlock, nil
+	}
+
+	if syncState == externalapi.SyncStateAwaitingBlockBodies {
+		headerTips, err := bp.headerTipsStore.Tips(bp.databaseContext)
+		if err != nil {
+			return 0, err
+		}
+		selectedHeaderTip, err := bp.ghostdagManager.ChooseSelectedParent(headerTips...)
+		if err != nil {
+			return 0, err
+		}
+		if *selectedHeaderTip != *blockHash {
```

Comment:
This condition is redundant

Reply:
We spoke about this and decided to leave it as is.